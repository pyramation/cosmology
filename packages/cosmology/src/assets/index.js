import json from './assetlist.json';

export const assets = json.assets;
export * from './helpers';
// https://raw.githubusercontent.com/osmosis-labs/assetlists/main/osmosis-1/osmosis-1.assetlist.json
