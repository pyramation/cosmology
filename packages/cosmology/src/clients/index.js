export * from './osmosis';
export * from './cosmos';
export * from './validator';