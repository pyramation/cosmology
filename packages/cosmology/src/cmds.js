
import _create_mnemonic_wallet from './commands/create-mnemonic-wallet';
import _decrypt from './commands/decrypt';
import _delegate from './commands/delegate';
import _encrypt from './commands/encrypt';
import _get_balances from './commands/get-balances';
import _get_cosmos_balance from './commands/get-cosmos-balance';
import _get_latest_block from './commands/get-latest-block';
import _get_pool_apr from './commands/get-pool-apr';
import _get_price from './commands/get-price';
import _get_staking_rewards from './commands/get-staking-rewards';
import _get_transaction from './commands/get-transaction';
import _ibc_transfer from './commands/ibc-transfer';
import _ibc_wip from './commands/ibc-wip';
import _join_pool from './commands/join-pool';
import _join from './commands/join';
import _list_apis from './commands/list-apis';
import _list_pools from './commands/list-pools';
import _list_prices from './commands/list-prices';
import _load_recipe from './commands/load-recipe';
import _lock_tokens from './commands/lock-tokens';
import _lock from './commands/lock';
import _reinvest_copy from './commands/reinvest-copy';
import _reinvest from './commands/reinvest';
import _salt_decrypt from './commands/salt-decrypt';
import _salt_encrypt from './commands/salt-encrypt';
import _salt_generate from './commands/salt-generate';
import _send from './commands/send';
import _shamir_decrypt from './commands/shamir-decrypt';
import _shamir_encrypt from './commands/shamir-encrypt';
import _swap_exact_amount_in from './commands/swap-exact-amount-in';
import _swap from './commands/swap';
import _wallet_from_mnemonic from './commands/wallet-from-mnemonic';
const Commands = {};
Commands['create-mnemonic-wallet'] = _create_mnemonic_wallet;
Commands['decrypt'] = _decrypt;
Commands['delegate'] = _delegate;
Commands['encrypt'] = _encrypt;
Commands['get-balances'] = _get_balances;
Commands['get-cosmos-balance'] = _get_cosmos_balance;
Commands['get-latest-block'] = _get_latest_block;
Commands['get-pool-apr'] = _get_pool_apr;
Commands['get-price'] = _get_price;
Commands['get-staking-rewards'] = _get_staking_rewards;
Commands['get-transaction'] = _get_transaction;
Commands['ibc-transfer'] = _ibc_transfer;
Commands['ibc-wip'] = _ibc_wip;
Commands['join-pool'] = _join_pool;
Commands['join'] = _join;
Commands['list-apis'] = _list_apis;
Commands['list-pools'] = _list_pools;
Commands['list-prices'] = _list_prices;
Commands['load-recipe'] = _load_recipe;
Commands['lock-tokens'] = _lock_tokens;
Commands['lock'] = _lock;
Commands['reinvest-copy'] = _reinvest_copy;
Commands['reinvest'] = _reinvest;
Commands['salt-decrypt'] = _salt_decrypt;
Commands['salt-encrypt'] = _salt_encrypt;
Commands['salt-generate'] = _salt_generate;
Commands['send'] = _send;
Commands['shamir-decrypt'] = _shamir_decrypt;
Commands['shamir-encrypt'] = _shamir_encrypt;
Commands['swap-exact-amount-in'] = _swap_exact_amount_in;
Commands['swap'] = _swap;
Commands['wallet-from-mnemonic'] = _wallet_from_mnemonic;

  export { Commands }; 

  