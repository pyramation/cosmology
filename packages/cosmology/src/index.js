export * from './assets';
export * from './clients';
export * from './utils/osmo';
export * from './utils/chain';
export * from './messages';
