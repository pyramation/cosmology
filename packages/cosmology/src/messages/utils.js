import { AminoTypes, SigningStargateClient } from '@cosmjs/stargate';
import { Registry } from '@cosmjs/proto-signing';
import { TxRaw } from 'cosmjs-types/cosmos/tx/v1beta1/tx';
import { coins } from '@cosmjs/amino';
import { defaultRegistryTypes } from '@cosmjs/stargate';
import retry from 'retry';
import { aminos } from './aminos';
import { meta as metaInfo } from './meta';

export const getSigningOsmosisClient = async ({ rpcEndpoint, signer }) => {
  // registry
  const registry = new Registry(defaultRegistryTypes);

  // aminotypes
  const aminoTypes = new AminoTypes({
    additions: Object.keys(aminos).reduce((m, key) => {
      const meta = metaInfo[key];
      const { toAmino, fromAmino } = aminos[key];
      m[meta.amino] = {
        aminoType: meta.type,
        toAmino,
        fromAmino
      };
      return m;
    }, {})
  });

  // register the goods
  Object.keys(aminos).forEach((key) => {
    const meta = metaInfo[key];
    registry.register(meta.amino, meta.osmosis);
  });

  const client = await SigningStargateClient.connectWithSigner(
    rpcEndpoint,
    signer,
    { registry: registry, aminoTypes: aminoTypes }
  );

  return client;
};

export const signAndBroadcast = async ({
  client,
  chainId,
  address,
  msg,
  fee,
  memo = ''
}) => {
  const { accountNumber, sequence } = await client.getSequence(address);
  const txRaw = await client.sign(address, [msg], fee, memo, {
    accountNumber: accountNumber,
    sequence: sequence,
    chainId
  });
  const txBytes = TxRaw.encode(txRaw).finish();
  return await client.broadcastTx(txBytes);
};

function getCosmosTx({ cosmos, transactionHash }) {
  const operation = retry.operation({
    retries: 5,
    factor: 2,
    minTimeout: 1 * 1000,
    maxTimeout: 60 * 1000
  });

  return new Promise((resolve, reject) => {
    operation.attempt(async () => {
      let results;
      let err;
      try {
        results = await cosmos.getCosmosTransaction(transactionHash);
      } catch (e) {
        console.log(e);
        err = true;
      }

      if (operation.retry(err)) {
        return;
      }

      // perhaps a tautology but semantics are that we
      // want to ensure tx is in the chain
      if (results.tx_response.txhash === transactionHash) {
        if (results.tx_response.code == 0) {
          resolve(results);
        } else {
          reject(results.tx_response.raw_log);
        }
      } else {
        reject(operation.mainError());
      }
    });
  });
}

export const signAndBroadcastTilTxExists = async ({
  client,
  cosmos,
  chainId,
  address,
  msg,
  fee,
  memo = ''
}) => {
  const result = await signAndBroadcast({
    client,
    chainId,
    address,
    msg,
    fee,
    memo
  });

  if (result.transactionHash) {
    const results = await getCosmosTx({
      cosmos,
      transactionHash: result.transactionHash
    });
    return results;
  }

  console.log(result);
  throw new Error('no tx hash');
};

export const generateOsmoMessage = (name, msg) => {
  if (!metaInfo[name]) throw new Error('missing message.');
  const gas = metaInfo[name].gas + ''; // TEST if needs string or if number is ok
  const fee = {
    amount: coins(0, 'uosmo'),
    gas
  };

  return {
    fee,
    msg: {
      typeUrl: metaInfo[name].amino,
      value: msg
    }
  };
};
