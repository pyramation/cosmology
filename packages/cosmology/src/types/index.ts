
export interface Pool {
    id: string;
    name: string;
    address: string;
    displayPoolAssets: object[];
    poolAssets: object[];
    pricePerShareEn18: string;
    totalShares: Coin;
    totalValue: string;
    totalWeight: string;
}

export interface TradeRoute {
    poolId: string;
    tokenOutDenom: CoinDenom;
    tokenOutSymbol: CoinSymbol;
    tokenInSymbol: CoinSymbol;
    liquidity: number;
}

export interface Swap {
    trade: Trade;
    routes: TradeRoute[];
};

export interface Pair {
    pool_address: string;
    pool_id: string;
    base_name: string;
    base_symbol: CoinSymbol;
    base_address: CoinDenom;
    quote_name: string;
    quote_symbol: CoinSymbol;
    quote_address: CoinDenom;
    price: number;
    base_volume_24h: number;
    quote_volume_24h: number;
    volume_24h: number;
    volume_7d: number;
    liquidity: number;
    liquidity_atom: number;
}
export interface LockedPoolDisplay {
    amount: string;
    denom: CoinDenom;
    value: string;
    allocation: string;
    poolId: string;
}
export interface Coin {
    amount: string;
    denom: CoinDenom;
}
export interface CoinWeight {
    weight: string;
    type: 'coin' | 'pool';
    name: string;
    value: string | null;
    symbol: CoinSymbol | null;
    poolId: string | null;
    denom: CoinDenom;
    allocation: string;
}
export interface CoinValue {
    amount: string;
    denom: CoinDenom;
    displayAmount: string;
    value: string;
    symbol: CoinSymbol;
}
export interface PoolAllocation {
    name: string;
    denom: CoinDenom;
    amount: string | null;
    displayAmount: string | null;
    value: string;
    coins: CoinValue[];
}
export interface DisplayCoin {
    amount: string;
    symbol: CoinSymbol;
}
export interface Trade {
    sell: CoinValue
    buy: CoinValue;
    beliefValue: string;
}
export type CoinDenom = string;

export type CoinSymbol =
    'ATOM' |
    'OSMO' |
    'ION' |
    'AKT' |
    'DVPN' |
    'IRIS' |
    'CRO' |
    'XPRT' |
    'REGEN' |
    'IOV' |
    'NGM' |
    'EEUR' |
    'JUNO' |
    'LIKE' |
    'UST' |
    'LUNA' |
    'BCNA' |
    'SCRT' |
    'MED';

export type CoinGeckoToken =
    'cosmos' |
    'osmosis' |
    'ion' |
    'akash-network' |
    'sentinel' |
    'iris-network' |
    'crypto-com-chain' |
    'persistence' |
    'regen' |
    'starname' |
    'e-money' |
    'e-money-eur' |
    'juno-network' |
    'likecoin' |
    'terrausd' |
    'terra-luna' |
    'bitcanna' |
    'terra-krw' |
    'secret' |
    'medibloc' |
    'comdex' |
    'cheqd-network' |
    'vidulum';

export interface PriceHash {
    [key: CoinDenom]: number;
}

export interface CoinGeckoUSD {
    usd: number;
};

export type CoinGeckoUSDResponse = Record<CoinGeckoToken, CoinGeckoUSD>


export interface ValidatorToken {
    price: number;
    denom: CoinDenom;
    symbol: CoinSymbol;
    liquidity: number;
    liquidity_24h_change: number;
    volume_24h: number;
    volume_24h_change: number;
    name: string;
};

export interface ValidatorTokenPrice {
    price: number;
}

export interface ValidatorCoinApr {
    start_date: string;
    denom: CoinDenom;
    symbol: CoinSymbol;
    apr_1d: number;
    apr_7d: number;
    apr_14d: number;
}

export interface ValidatorPoolApr {
    pool_id: number;
    apr_list: ValidatorCoinApr[]
}

export interface ValidatorPair {
    pool_address: string;
    pool_id: string;
    base_name: string;
    base_symbol: CoinSymbol;
    base_address: CoinDenom;
    quote_name: string;
    quote_symbol: CoinSymbol;
    quote_address: CoinDenom;
    price: number;
    base_volume_24h: number;
    quote_volume_24h: number;
    volume_24h: number;
    volume_7d: number;
    liquidity: number;
    liquidity_atom: number;
}

