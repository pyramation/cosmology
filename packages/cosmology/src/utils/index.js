export * from './chain';
export * from './helpers';
export * from './osmo';
export * from './print';
export * from './prompt';
export * from './prompts';
export * from './wallet';
export * from './gas';
