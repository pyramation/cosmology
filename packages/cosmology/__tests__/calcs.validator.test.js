// @ts-nocheck
import validatorPricesFixture from '../__fixtures__/validator/tokens/v2/all/data.json';
import poolsFixture from '../__fixtures__/lcd/osmosis/gamm/v1beta1/pools/data.json';
import lockedPoolsFixture from '../__fixtures__/lcd/osmosis/lockup/v1beta1/account_locked_coins/osmo1/data.json';
import cases from 'jest-in-case';
import { displayUnitsToDenomUnits } from '../src/utils';
import {
  osmoDenomToSymbol,
  symbolToOsmoDenom,
  convertCoinToDisplayValues,
  calculateCoinsTotalBalance,
  convertValidatorPricesToDenomPriceHash,
  getPoolByGammName,
  getUserPools,
  getFilteredPoolsWithValues,
  symbolsAndDisplayValuesToCoinsArray,
  getTradesRequiredToGetBalances
} from '../src/utils/osmo';

cases(
  'displayUnitsToDenomUnits',
  (opts) => {
    expect(displayUnitsToDenomUnits(opts.name, opts.amount)).toEqual(
      opts.value
    );
  },
  [
    { name: 'ATOM', amount: 10, value: '10000000.0000000000000000' },
    { name: 'AKT', amount: 0.6, value: '600000.0000000000000000' },
    { name: 'OSMO', amount: 10, value: '10000000.0000000000000000' }
  ]
);

cases(
  'osmoDenomToSymbol',
  (opts) => {
    expect(osmoDenomToSymbol(opts.denom)).toBe(opts.name);
  },
  [
    {
      name: 'ATOM',
      denom:
        'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2'
    },
    {
      name: 'AKT',
      denom:
        'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4'
    },
    { name: 'OSMO', denom: 'uosmo' }
  ]
);

cases(
  'symbolToOsmoDenom',
  (opts) => {
    expect(symbolToOsmoDenom(opts.name)).toBe(opts.denom);
  },
  [
    {
      name: 'ATOM',
      denom:
        'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2'
    },
    {
      name: 'AKT',
      denom:
        'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4'
    },
    { name: 'OSMO', denom: 'uosmo' }
  ]
);

cases(
  'convertValidatorPricesToDenomPriceHash',
  (opts) => {
    expect(convertValidatorPricesToDenomPriceHash(opts.tokens)).toEqual(
      opts.hash
    );
  },
  [
    {
      name: 'test1',
      tokens: [
        {
          denom:
            'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4',
          price: 100
        },
        {
          denom:
            'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2',
          price: 20
        }
      ],
      hash: {
        'ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4': 100,
        'ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2': 20
      }
    }
  ]
);

cases(
  'getPoolByGammName',
  (opts) => {
    // const prices = convertValidatorPricesToDenomPriceHash(validatorPricesFixture)
    const prices = convertValidatorPricesToDenomPriceHash(
      validatorPricesFixture
    );
    const pools = getFilteredPoolsWithValues({
      prices,
      pools: poolsFixture.pools
    });
    expect(getPoolByGammName(pools, opts.name).id).toEqual(opts.poolId);
  },
  [
    {
      name: 'gamm/pool/1',
      poolId: '1'
    },
    {
      name: 'gamm/pool/600',
      poolId: '600'
    },
    {
      name: 'gamm/pool/606',
      poolId: '606'
    }
  ]
);

describe('basic portfolio', () => {
  const coins = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'ATOM',
      amount: 10
    },
    {
      symbol: 'AKT',
      amount: 1000
    },
    {
      symbol: 'OSMO',
      amount: 100
    }
  ]);

  const prices = convertValidatorPricesToDenomPriceHash(validatorPricesFixture);

  it('calculate portfolio value (balances)', () => {
    const value = calculateCoinsTotalBalance({ prices, coins });
    expect(value).toBe('2622.486070510000000000');
  });

  it('calculate pool value (global)', () => {
    const values = getFilteredPoolsWithValues({
      prices,
      pools: poolsFixture.pools
    });
    const [first, second] = values;
    expect(first).toMatchSnapshot();
    expect(second).toMatchSnapshot();
  });

  it('calculate pool value (user LP)', () => {
    const pools = getFilteredPoolsWithValues({
      prices,
      pools: poolsFixture.pools
    });
    const result = getUserPools({
      pools,
      lockedPools: lockedPoolsFixture.coins
    });
    expect(result).toMatchSnapshot();
  });
});

describe('user actions', () => {
  const balances = symbolsAndDisplayValuesToCoinsArray([
    {
      symbol: 'ATOM',
      amount: 10
    },
    {
      symbol: 'AKT',
      amount: 1000
    },
    {
      symbol: 'OSMO',
      amount: 100
    }
  ]);

  const prices = convertValidatorPricesToDenomPriceHash(validatorPricesFixture);

  it('cannot desire more value than current', async () => {
    const desired = symbolsAndDisplayValuesToCoinsArray([
      {
        symbol: 'ATOM',
        amount: 100000
      }
    ]);
    let failed = false;
    try {
      getTradesRequiredToGetBalances({ prices, balances, desired });
    } catch (e) {
      failed = true;
    }
    expect(failed).toBe(true);
  });

  it('no trades required', async () => {
    const desired = symbolsAndDisplayValuesToCoinsArray([
      {
        symbol: 'ATOM',
        amount: 10
      },
      {
        symbol: 'OSMO',
        amount: 10
      }
    ]);
    const trades = getTradesRequiredToGetBalances({
      prices,
      balances,
      desired
    });
    expect(trades.length).toBe(0);
  });

  it('trades required', async () => {
    const desired = symbolsAndDisplayValuesToCoinsArray([
      {
        symbol: 'AKT',
        amount: 1050
      },
      {
        symbol: 'OSMO',
        amount: 100
      }
    ]);
    const trades = getTradesRequiredToGetBalances({
      prices,
      balances,
      desired
    });
    console.log(JSON.stringify(trades, null, 2));
  });
});

it('convertCoinToDisplayValues', async () => {
  const prices = convertValidatorPricesToDenomPriceHash(validatorPricesFixture);
  const a = convertCoinToDisplayValues({
    prices,
    coin: {
      denom: 'uosmo',
      amount: '10000'
    }
  });
  console.log(a);
});
