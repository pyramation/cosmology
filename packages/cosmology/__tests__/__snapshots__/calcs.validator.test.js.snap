// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic portfolio calculate pool value (global) 1`] = `
Object {
  "@type": "/osmosis.gamm.v1beta1.Pool",
  "address": "osmo1mw0ac6rwlp5r8wapwk3zs6g29h8fcscxqakdzw9emkne6c8wjp9q0t3v8t",
  "displayPoolAssets": Array [
    Object {
      "allocation": 0.5,
      "symbol": "ATOM",
      "token": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
      },
      "value": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
        "displayAmount": 6196127.370623,
        "symbol": "ATOM",
        "value": 177704932.98946765,
      },
      "weight": "536870912000000",
    },
    Object {
      "allocation": 0.5,
      "symbol": "OSMO",
      "token": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
      },
      "value": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
        "displayAmount": 24164665.57594,
        "symbol": "OSMO",
        "value": 192274877.97307968,
      },
      "weight": "536870912000000",
    },
  ],
  "future_pool_governor": "24h",
  "id": "1",
  "name": "ATOM/OSMO",
  "poolAssets": Array [
    Object {
      "token": Object {
        "amount": "6196127370623",
        "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
      },
      "weight": "536870912000000",
    },
    Object {
      "token": Object {
        "amount": "24164665575940",
        "denom": "uosmo",
      },
      "weight": "536870912000000",
    },
  ],
  "poolParams": Object {
    "exitFee": "0.000000000000000000",
    "smoothWeightChangeParams": null,
    "swapFee": "0.003000000000000000",
  },
  "pricePerShare": 1.120374312276983e-18,
  "totalShares": Object {
    "amount": "330228752041469100486167962",
    "denom": "gamm/pool/1",
  },
  "totalValue": 369979810.9625473,
  "totalWeight": "1073741824000000",
}
`;

exports[`basic portfolio calculate pool value (global) 2`] = `
Object {
  "@type": "/osmosis.gamm.v1beta1.Pool",
  "address": "osmo1500hy75krs9e8t50aav6fahk8sxhajn9ctp40qwvvn8tcprkk6wszun4a5",
  "displayPoolAssets": Array [
    Object {
      "allocation": 0.8,
      "symbol": "ION",
      "token": Object {
        "amount": "1289632142",
        "denom": "uion",
      },
      "value": Object {
        "amount": "1289632142",
        "denom": "uion",
        "displayAmount": 1289.632142,
        "symbol": "ION",
        "value": 9242398.48872225,
      },
      "weight": "858993459200000",
    },
    Object {
      "allocation": 0.2,
      "symbol": "OSMO",
      "token": Object {
        "amount": "374852643996",
        "denom": "uosmo",
      },
      "value": Object {
        "amount": "374852643996",
        "denom": "uosmo",
        "displayAmount": 374852.643996,
        "symbol": "OSMO",
        "value": 2982650.273214612,
      },
      "weight": "214748364800000",
    },
  ],
  "future_pool_governor": "24h",
  "id": "2",
  "name": "ION/OSMO",
  "poolAssets": Array [
    Object {
      "token": Object {
        "amount": "1289632142",
        "denom": "uion",
      },
      "weight": "858993459200000",
    },
    Object {
      "token": Object {
        "amount": "374852643996",
        "denom": "uosmo",
      },
      "weight": "214748364800000",
    },
  ],
  "poolParams": Object {
    "exitFee": "0.000000000000000000",
    "smoothWeightChangeParams": null,
    "swapFee": "0.005000000000000000",
  },
  "pricePerShare": 2.1147512230024997e-20,
  "totalShares": Object {
    "amount": "578084487147375945198234662",
    "denom": "gamm/pool/2",
  },
  "totalValue": 12225048.761936862,
  "totalWeight": "1073741824000000",
}
`;

exports[`basic portfolio calculate pool value (user LP) 1`] = `
Array [
  Object {
    "allocation": 0.0006056407831347302,
    "amount": "200000000000000000000000",
    "denom": "gamm/pool/1",
    "poolId": "1",
    "value": 224074.8624553966,
  },
  Object {
    "allocation": 0.00025312875703865847,
    "amount": "30000000000000000000000",
    "denom": "gamm/pool/3",
    "poolId": "3",
    "value": 3735.4518242542576,
  },
  Object {
    "allocation": 0.0001296341180589231,
    "amount": "200000000000000000000",
    "denom": "gamm/pool/497",
    "poolId": "497",
    "value": 8357.493733468797,
  },
  Object {
    "allocation": 0.00019494931711096735,
    "amount": "200000000000000000000",
    "denom": "gamm/pool/498",
    "poolId": "498",
    "value": 6394.8826621501885,
  },
  Object {
    "allocation": 0.0008236483020854653,
    "amount": "2000000000000000000000",
    "denom": "gamm/pool/560",
    "poolId": "560",
    "value": 127579.51927999774,
  },
  Object {
    "allocation": 0.0015710988409629195,
    "amount": "2000000000000000000000",
    "denom": "gamm/pool/561",
    "poolId": "561",
    "value": 129236.2175035652,
  },
  Object {
    "allocation": 0.0008591506333502901,
    "amount": "10000000000000000000000",
    "denom": "gamm/pool/562",
    "poolId": "562",
    "value": 68413.64019423944,
  },
  Object {
    "allocation": 0.0031545274576644427,
    "amount": "40000000000000000000",
    "denom": "gamm/pool/574",
    "poolId": "574",
    "value": 5325.237935985967,
  },
  Object {
    "allocation": 0.00011211369545863991,
    "amount": "20000000000000000000",
    "denom": "gamm/pool/584",
    "poolId": "584",
    "value": 2340.4835965732154,
  },
  Object {
    "allocation": 0.00028497385134167483,
    "amount": "400000000000000000",
    "denom": "gamm/pool/586",
    "poolId": "586",
    "value": 959.4257497075843,
  },
  Object {
    "allocation": 0.0018810841038095674,
    "amount": "30000000000000000000",
    "denom": "gamm/pool/592",
    "poolId": "592",
    "value": 3323.7097288922414,
  },
  Object {
    "allocation": 0.0014516082171006953,
    "amount": "40000000000000000000000000",
    "denom": "gamm/pool/6",
    "poolId": "6",
    "value": 12963.451454376986,
  },
  Object {
    "allocation": 0.004636281048913086,
    "amount": "500000000000000000",
    "denom": "gamm/pool/600",
    "poolId": "600",
    "value": 70745.98621455462,
  },
  Object {
    "allocation": 0.003314039294840055,
    "amount": "200000000000000000000",
    "denom": "gamm/pool/601",
    "poolId": "601",
    "value": 73243.34100968842,
  },
  Object {
    "allocation": 0.0027907320014314866,
    "amount": "400000000000000000",
    "denom": "gamm/pool/604",
    "poolId": "604",
    "value": 65906.21452946201,
  },
  Object {
    "allocation": 0.0008335361960606539,
    "amount": "300000000000000000000000",
    "denom": "gamm/pool/606",
    "poolId": "606",
    "value": 4059.264921460508,
  },
  Object {
    "allocation": 0.002495331745914565,
    "amount": "4000000000000000000000",
    "denom": "gamm/pool/611",
    "poolId": "611",
    "value": 50179.95986579703,
  },
  Object {
    "allocation": 0.0003611517731477527,
    "amount": "600000000000000000000000",
    "denom": "gamm/pool/9",
    "poolId": "9",
    "value": 7395.407933505083,
  },
]
`;
